test/powershell/Modules/Microsoft.PowerShell.Core/Enter-PSHostProcess.Tests.ps1

```diff
@@ -1,134 +1,163 @@
+    } while (!$pwshId)
+
+    $pwshId
+}
+
+# Executes the Enter/Exit PSHostProcess script that returns the pid of the process that's started.
+function Invoke-PSHostProcessScript {
+    param (
+        [string] $ArgumentString,
+        [int] $Id,
+        [int] $Retry = 5 # Default retry of 5 times
+    )
+
+    $sb = {
+        $commandStr = @'
+Start-Sleep -Seconds {0}
+Enter-PSHostProcess {1} -ErrorAction Stop
+$pid
+Exit-PSHostProcess
+'@ -f $Retry, $ArgumentString
+
+        ($commandStr | pwsh -c -) -eq $Id
+    }
+
+    $result = $false
+    foreach ($i in 1..$Retry) {
+        if ($sb.Invoke()) {
+            $result = $true
+            break
+        }
+
+        Write-Warning "Enter-PSHostProcess attempt '$i' failed. $($Retry - $i) attempts left."
+    }
+
+    $result
+}
+
 Describe "Enter-PSHostProcess tests" -Tag Feature {
     Context "By Process Id" {
-        BeforeAll {
-            $params = @{
-                FilePath = "pwsh"
-                PassThru = $true
-                RedirectStandardOutput = "TestDrive:\pwsh_out.log"
-                RedirectStandardError = "TestDrive:\pwsh_err.log"
-            }
-            $pwsh = Start-Process @params
-
-            if ($IsWindows) {
-                $params = @{
-                    FilePath = "powershell"
-                    PassThru = $true
-                    RedirectStandardOutput = "TestDrive:\powershell_out.log"
-                    RedirectStandardError = "TestDrive:\powershell_err.log"
+
+        BeforeEach {
+            # Start a normal job where the first thing it does is return $pid. After that, spin forever.
+            # We will use this job as the target process for Enter-PSHostProcess
+            $pwshJob = Start-Job {
+                $pid
+                while ($true) {
+                    Start-Sleep -Seconds 30 | Out-Null
                 }
-                $powershell = Start-Process @params
             }
 
+            $pwshId = Wait-JobPid $pwshJob
         }
 
-        AfterAll {
-            $pwsh | Stop-Process
-
-            if ($IsWindows) {
-                $powershell | Stop-Process
-            }
+        AfterEach {
+            $pwshJob | Stop-Job -PassThru | Remove-Job
         }
 
-        It "Can enter and exit another PSHost" -Pending:$true {
-            Wait-UntilTrue { (Get-PSHostProcessInfo -Id $pwsh.Id) -ne $null }
+        It "Can enter, exit, and re-enter another PSHost" {
+            Wait-UntilTrue { [bool](Get-PSHostProcessInfo -Id $pwshId) }
 
-            "Enter-PSHostProcess -Id $($pwsh.Id) -ErrorAction Stop
-            `$pid
-            Exit-PSHostProcess" | pwsh -c - | Should -Be $pwsh.Id
+            # This will enter and exit another process
+            Invoke-PSHostProcessScript -ArgumentString "-Id $pwshId" -Id $pwshId |
+                Should -BeTrue -Because "The script was able to enter another process and grab the pid of '$pwshId'."
+
+            # Re-enter and exit the other process
```

Comment:
I feel re-enter functionality does not need to be part of our automated tests, and can be removed.

Reply:
It tests whether or not the NamedPipeServer properly gets restarted - something that wouldn't be tested if this code wasn't in the test. I think that's a pretty valid reason for a test, no?

Reply:
My concern is test fragility.  Some tests are difficult to make reliable and when they fail regularly cause a lot of trouble and confusion during build and release.  In these cases I feel it is best to test basic functionality and not all edge cases.  So it seems interactive cmdlets and re-connect tests can be left out of our automated tests.

Reply:
OTOH if these tests run reliably we should leave them in.  Let's see how they do before we decide.